samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -288,22 +300,43 @@ void BLEConnectivityManager::DiscoveryCompletedHandler(bt_gatt_dm *dm, void *con
 
 	discoveryResult = true;
 exit:
-	if (provider) {
-		if (!provider->IsInitiallyConnected()) {
-			/* Provider is not initalized it so we need to call the first connection callback */
-			provider->GetBLEBridgedDevice().mFirstConnectionCallback(
-				discoveryResult, provider->GetBLEBridgedDevice().mFirstConnectionCallbackContext);
-			provider->ConfirmInitialConnection();
-		}
-
-		VerifyOrReturn(CHIP_NO_ERROR == provider->NotifyReachableStatusChange(true),
-			       LOG_WRN("The device has not been notified about the status change."));
 
-		VerifyOrReturn(provider->ParseDiscoveredData(dm) == 0,
-			       LOG_ERR("Cannot parse the GATT discovered data."));
+	Platform::UniquePtr<DiscoveryHandlerCtx> discoveryCtx(Platform::New<DiscoveryHandlerCtx>());
+	if (!discoveryCtx) {
+		return;
 	}
 
-	bt_gatt_dm_data_release(dm);
+	discoveryCtx->mProvider = provider;
+	discoveryCtx->mDiscoveryData = dm;
+
+	CHIP_ERROR err = chip::DeviceLayer::PlatformMgr().ScheduleWork(
+		[](intptr_t context) {
+			auto *ctx = reinterpret_cast<DiscoveryHandlerCtx *>(context);
+			if (!ctx->mProvider->IsInitiallyConnected()) {
+				/* Provider is not initalized it so we need to call the first connection callback */
+				CHIP_ERROR err = ctx->mProvider->GetBLEBridgedDevice().mFirstConnectionCallback(
+					ctx->mDiscoveryData,
+					ctx->mProvider->GetBLEBridgedDevice().mFirstConnectionCallbackContext);
+				ctx->mProvider->ConfirmInitialConnection();
+				VerifyOrReturn(CHIP_NO_ERROR == err, bt_gatt_dm_data_release(ctx->mDiscoveryData););
+			}
+
+			VerifyOrReturn(CHIP_NO_ERROR == ctx->mProvider->NotifyReachableStatusChange(true),
+				       LOG_WRN("The device has not been notified about the status change."));
+
+			VerifyOrReturn(ctx->mProvider->ParseDiscoveredData(ctx->mDiscoveryData) == 0,
```

Comment:
Ditto, I think we should always release gatt data and delete context, no matter if we failed or not.